File: leetcode/131_Palindrome_Partitioning.py
from typing import List
import unittest

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def _rec(start: int, partition: List[str], partitions: List[List[str]]) -> None:
            if start >= n:
                partitions.append(partition.copy())
            else:
                for end in range(start, n):
                    if dp[start][end]:
                        partition.append(s[start : end + 1])
                        _rec(end + 1, partition, partitions)
                        partition.pop()

        n, partition, partitions = len(s), [], []
        dp = [[False] * n for i in range(n)]
        for i in range(n):
            dp[i][i] = True
        for l in range(1, n):
            for i in range(0, n - l):
                if s[i] == s[i + l] and (l == 1 or dp[i + 1][i + l - 1]):
                    dp[i][i + l] = True
        _rec(0, partition, partitions)
        return partitions


class PartitionTest(unittest.TestCase):
    def test_1(self) -> None:
        self.assertEqual(Solution().partition("aab"), [["a","a","b"],["aa","b"]]) 

    def test_2(self) -> None:
        self.assertEqual(Solution().partition("a"), [["a"]]) 

File: leetcode/131_Palindrome_Partitioning.rs
struct Solution {}

impl Solution {
    pub fn partition(s: String) -> Vec<Vec<String>> {
        fn _rec(s: &Vec<char>, start: usize, n: usize, partition: &mut Vec<(usize, usize)>, partitions: &mut Vec<Vec<String>>, dp: &Vec<Vec<bool>>) {
            if start >= n {
                partitions.push(partition.iter().map(| (i, j) | s[*i .. *j + 1].into_iter().collect()).collect());
            } else {
                for end in start..n {
                    if dp[start][end] {
                        partition.push((start, end));
                        _rec(s, end + 1, n, partition, partitions, dp);
                        partition.pop();
                    }
                }
            }
        }
        let s: Vec<char> = s.chars().collect();
        let n = s.len();
        let mut partitions = vec![];
        let mut partition = vec![];
        let mut dp = vec![vec![false; n]; n];
        for i in 0..n {
            dp[i][i] = true;
        }
        for l in 1..n {
            for i in 0..n-l {
                if s[i] == s[i + l] && (l == 1 || dp[i + 1][i + l - 1]) {
                    dp[i][i + l] = true;
                }
            }
        }
        _rec(&s, 0, n, &mut partition, &mut partitions, &dp);
        partitions
    }
}

mod partition_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(
            Solution::partition("aab".to_string()),
            vec![vec!["a".to_string(),"a".to_string(),"b".to_string()], vec!["aa".to_string(),"b".to_string()]]
        );
    }    

    #[test]
    fn test_2() {
        assert_eq!(
            Solution::partition("a".to_string()),
            vec![vec!["a".to_string()]]
        );
    }    
}    